File: c++/src/Team4343/Subsystems/Chassis.cpp
#include "Chassis.h"
#include "../Robotmap.h"
#include "../Commands/DriveWithJoystick.h"
#include "../Defines.h"

Chassis::Chassis() : Subsystem("Chassis") {
	printf("Initializing chassis.\n");
	drive = new RobotDrive(LEFT_MOTOR, RIGHT_MOTOR);
	drive->SetSafetyEnabled(SAFETY_ENABLED);

	drive->SetExpiration(EXPIRATION);
}

void Chassis::InitDefaultCommand() {
	SetDefaultCommand(new DriveWithJoystick());
}

void Chassis::goStraight() {
	drive->ArcadeDrive(1.0, 0.0);
}

void Chassis::turn(double turnVal) {
	drive->ArcadeDrive(0.0, turnVal);
}

void Chassis::tankDrive(double left, double right) {
	drive->TankDrive(left, right);
}

void Chassis::driveWithJoystick(Joystick *stick, float previousXVal, float previousYVal) {
    float currentXVal = (stick->GetAxis(stick->kXAxis));
    float currentYVal = (stick->GetAxis(stick->kYAxis));

    cout << currentXVal;
    cout << currentYVal;

    float accelerationRate = 0.2;

    float combinedX = previousXVal + (currentXVal - previousXVal) * accelerationRate;
    float combinedY = previousYVal + (currentYVal - previousYVal) * accelerationRate;

    previousXVal = combinedX;
    previousYVal = combinedY;

    cout << combinedX;
    cout << combinedY;

    drive->ArcadeDrive(combinedX, combinedY);
}

void Chassis::driveWithTurn(double distance, double turnVal) {
	drive->ArcadeDrive(distance, turnVal);
}

bool Chassis::isMoving() {
	return drive->IsAlive();
}

File: java/src/com/frc4343/robot/subsystems/Chassis.java
package com.frc4343.robot.subsystems;

import com.frc4343.robot.Constants;
import com.frc4343.robot.Mappings;
import com.frc4343.robot.commands.DriveWithJoystick;
import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.RobotDrive;
import edu.wpi.first.wpilibj.command.Subsystem;

public class Chassis extends Subsystem {
    RobotDrive drive; // Used to drive the robot.
    double speed; // Holds the current speed of the robot.

    private double previousXVal = 0.0;
    private double previousYVal = 0.0;

    public Chassis() {
        super("Chassis");

        System.out.println("Initializing chassis.");

        drive = new RobotDrive(Mappings.LEFT_MOTOR, Mappings.RIGHT_MOTOR);

        drive.setSafetyEnabled(Constants.SAFETY_ENABLED);
        drive.setExpiration(Constants.EXPIRATION);
    }

    public void initDefaultCommand() {
        // Set the default command for a subsystem here.
        setDefaultCommand(new DriveWithJoystick());
    }

    public void goStraight() {
        drive.arcadeDrive(1.0, 0.0);
    }

    public void turn(double turnVal) {
        drive.arcadeDrive(0.0, turnVal);
    }

    public void tankDrive(double left, double right) {
        drive.tankDrive(left, right);
    }

    public void driveWithJoystick(Joystick stick) {
        double currentXVal = stick.getAxis(Joystick.AxisType.kY);
        double currentYVal = stick.getAxis(Joystick.AxisType.kX);

        double accelerationRate = 0.2;

        previousXVal = getCurrentXVal() + (currentXVal - getCurrentXVal()) * accelerationRate;
        previousYVal = getCurrentYVal() + (currentYVal - getCurrentYVal()) * accelerationRate;

        drive.arcadeDrive(getCurrentXVal(), getCurrentYVal());
    }

    public void driveWithTurn(double distance, double turnVal) {
        drive.arcadeDrive(distance, turnVal);
    }

    public boolean isMoving() {
        return drive.isAlive();
    }

    public double getCurrentXVal() {
        return previousXVal;
    }

    public double getCurrentYVal() {
        return previousYVal;
    }
}
